RavenDB-22142 - Schema Validation

grisha-kotler opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22142/Schema-Validation

### Additional description

## 🚀 Feature: Schema Validation per Collection

This PR introduces the ability to enforce a **JSON Schema** on documents before they are saved into a collection. This ensures data consistency and integrity, as any document that does not conform to the defined schema will be **rejected** upon the save operation.

---

### 🌟 Key Capabilities

* **Per-Collection Configuration:** Users can define a unique validation schema for each collection within a database.
* **Rejection of Invalid Documents:** If a document fails validation against the configured schema during a `PUT` or `PATCH` operation, the save operation is aborted, and an error is returned.
* **Industry Standard:** The feature utilizes the widely adopted **JSON Schema** standard ([https://json-schema.org/](https://json-schema.org/)) for defining validation rules.

---

### 🛠️ Implementation Details

#### 1. Server-Side Changes

* **Storage:** A new configuration object is persisted per collection to store its validation schema (the JSON Schema definition) and an activation flag.
* **Validation Hook:** A validation layer is introduced in the document-save pipeline. Before a document is persisted, it is checked against the configured schema for its collection.
* **Error Handling:** A specific exception type (`SchemaValidationException`) is thrown upon validation failure, including details about the schema violations.

#### 2. Client API Commands

To manage the schema validation settings, two new client-side operations have been added:

| Operation | Purpose |
| :--- | :--- |
| `ConfigureSchemaValidationOperation` | Sets or updates the JSON Schema and enables/disables validation for a specified collection. |
| `GetSchemaValidationConfiguration` | Retrieves the current validation schema and configuration status for a specified collection. |

---

### 💡 Example Usage (Conceptual)

Imagine a collection named `Products`. A user can define a schema that forces all documents to have a `Name` (string) and a `Price` (number, greater than 0).

**Schema Definition:**
```json
{
  "type": "object",
  "properties": {
    "Name": { "type": "string" },
    "Price": { "type": "number", "minimum": 0, "exclusiveMinimum": true }
  },
  "required": ["Name", "Price"],
  "additionalProperties": false
}
```

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

haludi committed:
RavenDB-22142 Base for schema validation with basic rules

haludi committed:
RavenDB-22142 Enum rule validator

haludi committed:
RavenDB-22142 Constant validator

haludi committed:
RavenDB-22142 Boolean validation

haludi committed:
RavenDB-22142 Validate null

haludi committed:
RavenDB-22142 Handle PatternProperties

haludi committed:
RavenDB-22142 Support additionalProperties

haludi committed:
RavenDB-22142 Fix path

haludi committed:
RavenDB-22142 Support restriction on nested object

haludi committed:
RavenDB-22142 Add errors for invalid schema

haludi committed:
RavenDB-22142 Move 'required' to its own validator

haludi committed:
RavenDB-22142 Change the tests to assert in parallel

haludi committed:
RavenDB-22142

haludi committed:
RavenDB-22142 Property Names Validator

haludi committed:
RavenDB-22142 Validate array type

haludi committed:
RavenDB-22142 Add uniqueItems and automatically find validators

haludi committed:
RavenDB-22142 Move rules name to const

haludi committed:
RavenDB-22142 Validate array

haludi committed:
RavenDB-22142 Allow to validate without error message building

haludi committed:
RavenDB-22142 Move Path to ErrorBuilder

haludi committed:
RavenDB-22142 Array contains rule validation

haludi committed:
RavenDB-22142 Move contains rule validation to its own validator

haludi committed:
RavenDB-22142 Add 'minContains' & 'maxContains'

haludi committed:
RavenDB-22142 dependentRequired

haludi committed:
RavenDB-22142 Conditional Schema

haludi committed:
RavenDB-22142 Move init code to factory

haludi committed:
RavenDB-22142 dependent schema

haludi committed:
RavenDB-22142 Schema path

haludi committed:
RavenDB-22142 Reduce error message allocation

haludi committed:
RavenDB-22142 RentedBuffer

haludi committed:
RavenDB-22142 `Not` validator

haludi committed:
RavenDB-22142 Reference Schema

haludi committed:
RavenDB-22142 Minor refactor SchemaRuleValidator definition

haludi committed:
RavenDB-22142 Unmanaged buffers for errors

haludi committed:
RavenDB-22142 Fix exceptions type in ErrorBufferStreamWrapper

haludi committed:
RavenDB-22142 Use direct value

haludi committed:
RavenDB-22142 Arrange untyped validators

haludi committed:
RavenDB-22142 Ref schema

haludi committed:
RavenDB-22142 AllOf validator

haludi committed:
RavenDB-22142 oneOf validator

haludi committed:
RavenDB-22142 anyOf validator

grisha-kotler committed:
RavenDB-24253 - add the ConfigureSchemaValidationOperation and GetSchemaValidationConfiguration

grisha-kotler committed:
RavenDB-24254 - connect the SchemaValidator to the DocumentsStorage

grisha-kotler committed:
RavenDB-24254 - update tests

grisha-kotler committed:
RavenDB-24254 - skip validation when the document comes from replication or resharding

grisha-kotler committed:
RavenDB-24254 - update endpoints

grisha-kotler committed:
RavenDB-24254 - add missing command to JsonDeserializationCluster

haludi committed:
RavenDB-22142 Avoid using context concurrently when comparing, fix LazyStringValue.TryFormat, remove unnecessary scopes from tests

grisha-kotler committed:
RavenDB-24254 - update DatabaseRecordItemType

grisha-kotler committed:
RavenDB-24253 - update smuggler

grisha-kotler committed:
RavenDB-24254 - fix NRE during restore

grisha-kotler committed:
RavenDB-24254 - update cluster command version

grisha-kotler committed:
RavenDB-24254 - fix test

haludi committed:
RavenDB-22142 Use RavenFact

grisha-kotler committed:
RavenDB-24253 - set SchemaDefinition to string

grisha-kotler committed:
RavenDB-24253 - fix NRE

grisha-kotler committed:
RavenDB-24253 - add a basic test

grisha-kotler committed:
RavenDB-24253 - add support for additionalProperties - @metadata is a mandatory property in the document

grisha-kotler committed:
RavenDB-24254 - don't throw an exception if we failed to parse the schema

grisha-kotler committed:
RavenDB-24254 - add the SchemaValidationException

haludi committed:
RavenDB-24319 Add test to validation schema validation across data flow scenarios

haludi committed:
RavenDB-24319 Modify restore, export/import database record test to include schema validation

haludi committed:
RavenDB-24319 Fix tests

haludi committed:
RavenDB-22142 Explicit case-sensitive for schema references

haludi committed:
RavenDB-22142 Use FrozenDictionary for read-only validator dictionary

haludi committed:
RavenDB-22142 Properly dispose buffers

haludi committed:
RavenDB-22142 HashSet, fix names and scopes, AdditionalProperties & requiredProperties type check

haludi committed:
RavenDB-22142 Move the SchemaValidation folder inside the Documents folder

haludi committed:
RavenDB-22142 Back property for ValidatorsPerCollection

ppekrol committed:
Merge pull request #21181 from ppekrol/feature/json-schema

7.1 to feature/json-schema merge

haludi committed:
RavenDB-22142 Fix tests

ppekrol committed:
Merge branch 'v7.1' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #21222 from ppekrol/feature/json-schema

merge 7.1 to feature/json-schema

haludi committed:
RavenDB-22142 Remove unnecessary Flush and SchemaValidator -> SchemaDefinition

ppekrol committed:
Merge branch 'v7.1' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #21342 from ppekrol/feature/json-schema

Merge 7.1 into feature/json-schema

M4xymm committed:
RavenDB-24346 feat (document schema) - add Document Schema settings page and integrate with navigation and app URLs

RavenDB-24346 feat (document schema) - Added DocumentSchema.tsx component as initial work without business logic, added stories

RavenDB-24346 feat (document schema) - add Document Schema Playground page and integrate with app URLs and navigation

RavenDB-24346 feat (document schema) - Added layout DocumentSchemaPlayground.tsx.

M4xymm committed:
RavenDB-24346 feat (document schema) - implement document schema management with actions, selectors, utils, and UI integration.

M4xymm committed:
RavenDB-24346 feat (document schema) - restructure components into partials, extract `useDocumentSchema` hook, and temporarily disable Document Schema Playground.

M4xymm committed:
RavenDB-24346 refactor (document schema) - remove unused `license` variable from mock store in DocumentSchema stories

M4xymm committed:
RavenDB-24346 feat (document schema) - integrate `setValue` in `AceEditor.LoadFileAction` to update schema on file load

M4xymm committed:
RavenDB-24346 feat (document schema) - add unit tests for DocumentSchema component, improve JSON validation, and integrate `useWatch` for dynamic schema updates

haludi committed:
RavenDB-24485 Allow to index schema validity

ppekrol committed:
Merge branch 'v7.1' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #21466 from ppekrol/feature/json-schema

7.1 to feature/json-schema merge

kalczur committed:
RavenDB-24346 Fix AceEditor onValidate

M4xymm committed:
RavenDB-24346 feat (document schema) - Fixed Cr comments

M4xymm committed:
RavenDB-24346 feat (document schema) - Fixed tests

M4xymm committed:
RavenDB-24346 feat (document schema) - eslint fix

mateuszbartosik committed:
RavenDB-24346 Add document-schema icon

M4xymm committed:
RavenDB-24346 feat (document schema) - fixed getSchemaValidationCommand.ts

M4xymm committed:
RavenDB-24346 feat (document schema) - Fixed filtering, added more stubs to storybook

M4xymm committed:
RavenDB-24346 feat (document schema) - Fixed editing schema validation collection name

M4xymm committed:
RavenDB-24346 feat (document schema) - Added disable/enable to document schema validation

M4xymm committed:
RavenDB-24346 feat (document schema) - PRETTIER

haludi committed:
RavenDB-24485 Allow to get schema validation status for a collection

ppekrol committed:
Merge branch 'v7.1' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #21521 from ppekrol/feature/json-schema

Merge 7.1 into feature/json-schema

haludi committed:
RavenDB-24485 One schema validation for index, fix operation parameter check,

haludi committed:
RavenDB-24485 Throw when validate 'null' in index

Danielle9897 committed:
RDoc-3515 About text for Document Schema view + tooltips

haludi committed:
RavenDB-24485 `ValidateSchemaValidationOperation` propertiy names

M4xymm committed:
RavenDB-24346 feat (document schema) - Added filtering by status and refactored collection filter into a reusable component

M4xymm committed:
RavenDB-24346 feat (document schema) - Added global toggle for enabling/disabling schema validation

M4xymm committed:
RavenDB-24346 feat (document schema) - Improved UI and logic for enabling/disabling schema validation globally and per schema

haludi committed:
RavenDB-24485 Use Stopwatch instead of current time.

haludi committed:
RavenDB-24485 Parameters.SchemaDefinition name and comment

M4xymm committed:
RavenDB-24346 feat (document schema) - Added JSON Schema validation using Ajv library

haludi committed:
RavenDB-24485 Only one index function and support javascript

src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs

haludi committed:
RavenDB-24485 Exclude @metadata

haludi committed:
RavenDB-24485 Add etag to ValidateSchemaValidationOperation

ppekrol committed:
Merge branch 'v7.1' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #21577 from ppekrol/feature/json-schema

Merge 7.1 into feature/json-schema

M4xymm committed:
RavenDB-24346 feat (document schema) - Renamed `globalDisabled` to `isGlobalDisabled` across the codebase for clarity and consistency

haludi committed:
RavenDB-24485 Cleaning code

haludi committed:
RavenDB-24485 Avoid VlidateValidate in names

haludi committed:
RavenDB-24485 Concat `excludedProperties`

haludi committed:
RavenDB-24485 Move javascript schemaValidate func to schema object

	src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs

haludi committed:
RavenDB-24485 Explicitly convert string to ReadOnlySpan<char>

ppekrol committed:
Merge branch 'v7.1' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #21775 from ppekrol/feature/json-schema

merge 7.1 into feature/json-schema

haludi committed:
RavenDB-24485 Avoid "Validate..Validate" in names

	modified:   src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaResult.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Handlers/SchemaValidationHandler.cs
	modified:   src/Raven.Server/Documents/Indexes/Index.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/CurrentIndexingScope.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs
	modified:   src/Raven.Server/Documents/Operations/OperationType.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/ObjectSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/Sharding/Handlers/Processors/SchemaValidation/ShardedSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Sharding/Handlers/ShardedSchemaValidationHandler.cs
	modified:   src/Raven.Server/Json/JsonDeserializationServer.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs

haludi committed:
RavenDB-24485 Index array of errors and index configuration per collection

	modified:   src/Raven.Client/Documents/Indexes/IndexDefinition.cs
	new file:   src/Raven.Client/Documents/Indexes/IndexSchemaDefinitions.cs
	modified:   src/Raven.Server/Documents/Indexes/Index.cs
	modified:   src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs
	new file:   src/Raven.Server/Documents/Indexes/SchemaValidation/IndexSchemaValidatorsByCollection.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/CurrentIndexingScope.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/MapIndexDefinition.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/Roslyn/MapFunctionProcessor.cs
	new file:   src/Raven.Server/Documents/Indexes/Static/Roslyn/Rewriters/SchemaValidatorRewriter.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs
	modified:   src/Raven.Server/Documents/Patch/JavaScriptUtils.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/AbstractBuffer.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/ErrorBuffer.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/ErrorBuilder.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/IErrorBuffer.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ValidationPath.cs
	modified:   src/Raven.Server/Extensions/JsonSerializationExtensions.cs
	modified:   src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs
	modified:   src/Raven.Server/Json/Sync/BlittableJsonTextWriterExtensions.cs
	modified:   src/Raven.Server/ServerWide/ServerStore.cs
	modified:   test/FastTests/SchemaValidation/SchemaValidationTestsBase.cs
	modified:   test/SlowTests.Issues/Issues/RavenDB-17597.cs
	modified:   test/SlowTests/SchemaValidation/BasicIntegrationSchemaValidationTests.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs
	modified:   test/SlowTests/Server/Documents/Indexing/SchemaValidationIndexingTests.cs

haludi committed:
RavenDB-24485 Rename ValidateSchemaOperation to StartSchemaValidationOperation and update related references

	renamed:    src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaOperation.cs -> src/Raven.Client/Documents/Operations/SchemaValidation/StartSchemaValidationOperation.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Sharding/Handlers/Processors/SchemaValidation/ShardedSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Json/JsonDeserializationServer.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs

ArieSLV committed:
RavenDB-25382 - Handling the `SchemaValidation` license attribute

ArieSLV committed:
RavenDB-25382 - Move LicenseManager's AssertCanAddSchemaValidation to the proper place

ppekrol committed:
Merge branch 'v7.1' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #21865 from ppekrol/feature/json-schema

Merge 7.1 into feature/json-schema

ArieSLV committed:
RavenDB-25382 - PR Comment

ArieSLV committed:
RavenDB-25382 - Studio fix

ArieSLV committed:
RavenDB-25382 - Add LicenseRequired attribute to Store test in SchemaValidationBasicTests

haludi committed:
RavenDB-24485 Disallow revert revision when shema validation is enabled.

M4xymm committed:
RavenDB-25473 working notification for validate schema with grid

RavenDB-25473 (WIP) - feat (schema validation): Added validation schema view sheet panel with progress tracking and test settings configuration. Updated related operations and UI enhancements for schema validation.

RavenDB-25473 feat (document schema playground): Introduced improved schema validation with support for multiple validators, dynamic progress tracking, and enhanced UI components. Refactored unused code and added collapsible panels.

RavenDB-25473 feat (document schema validation): Refactored validation operations to enhance test configuration, progress monitoring, and error presentation. Introduced reusable components for validation status and user interaction. Added operation abort functionality and improved table column rendering.

grisha-kotler committed:
Merge branch 'v7.2' of https://github.com/ravendb/ravendb into v7.2

ppekrol committed:
Merge pull request #21923 from grisha-kotler/Schema-Validation-7.2

Merge 7.2 to Json Schema feature branch

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -446,6 +446,7 @@ record = _serverStore.Cluster.ReadDatabase(context, Name, out index);
 
                 ReplicationLoader = CreateReplicationLoader();
                 PeriodicBackupRunner = new PeriodicBackupRunner(this, _serverStore, wakeup);
+                SchemaValidatorCache = new SchemaValidatorCache(DocumentsStorage.ContextPool, Loggers.GetLogger<SchemaValidatorCache>());
```
I would rather this be `null` if we have no schema defined.
That way, it is a predicated branch in the `DocumentPutAction`, instead of a call + branch.

arekpalinski replied:
The code seems to be already prepared for this being `null` as it does `SchemaValidatorCache?.` wherever it is used.

grisha-kotler replied:
Addressed here: https://github.com/ravendb/ravendb/pull/21966

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs:
```diff
@@ -0,0 +1,145 @@
+                if (_logger.IsErrorEnabled)
+                    _logger.Error($"Failed to parse the schema validator for collection {collection}", e);
+
+                continue;
```
Just to make sure, it is just defensive coding, right? We do validate that upon receiving a configuration from a user, don't we?

haludi replied:
I added this in the second PR
https://github.com/ravendb/ravendb/pull/21511/files#diff-343189e415881abff6c312fa21edc9c5b752335a99df35ee03bf75419bb5fd60

I also added a notification
https://github.com/ravendb/ravendb/pull/21511/files#diff-cfdfbdc387ae5ddd4a353363e82a3da0c984d25a4027926f74d73f167aef5bd2R75

arekpalinski replied:
That's great

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs:
```diff
@@ -0,0 +1,145 @@
+
+    public void Validate(string collection, BlittableJsonReaderObject document, NonPersistentDocumentFlags nonPersistentFlags, DocumentsOperationContext context)
+    {
+        // TODO: check if we need to add more flags here
```
TODO?

haludi replied:
I will address it as part of 
https://issues.hibernatingrhinos.com/issue/RavenDB-25564

arekpalinski replied:
ok

arekpalinski commented on src/Raven.Client/Documents/Operations/SchemaValidation/SchemaValidationConfiguration.cs:
```diff
@@ -0,0 +1,37 @@
+{
+    private Dictionary<string, SchemaDefinition> _validatorsPerCollection;
+    
+    public bool Disabled { get; set; }
```
Is there any way to remove the schema validation? I found only this property to disable it

arekpalinski replied:
Do we miss `DeleteSchemaValidationOperation`?

haludi replied:
There is no API to add (or remove) a schema.
That means, no "patch" for the configuration.
Only "put" of the entire configuration.
So if you want to remove, you need to put it without the schema for the relevant collection.
When you refer to "delete" do you want a delete for the entire configuration, or do you refer to only one of a specific collection?

arekpalinski replied:
I was thinking about removing it completely - so about entire configuration

arekpalinski replied:
so I thought about having `DeleteSchemaValidationOperation`

grisha-kotler replied:
We were trying to match the behaviour of other commands that update the database record and also update the whole configuration (like we have here - we update the whole configuration and not by collection). For example: `EditDocumentsCompressionCommand`, `EditRevisionsConfigurationCommand`, `EditTimeSeriesConfigurationCommand`.

If you want to delete the configuration, you can put an empty one (or disable the entire configuration).

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidator.cs:
```diff
@@ -0,0 +1,47 @@
+        Disabled = disabled;
+    }
+    
+    // TODO: add a comment about the blittable lifetime
```
TODO?

haludi replied:
Already addressed in [21511](https://github.com/ravendb/ravendb/pull/21511)

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidator.cs:
```diff
@@ -0,0 +1,47 @@
+    public readonly bool Disabled;
+
+    private ElementSchemaRuleValidator _root;
+    //The context is only written during the initialization phase. During validation, it is used for reading only and can be used in parallel.
```
What 'context' are you referring to?

haludi replied:
The schema blittable context.
The blittalbe memory/data is used for nested blittalbe when validating "const"
and for `LazyStringValue`s.

arekpalinski replied:
ok

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidator.cs:
```diff
@@ -0,0 +1,47 @@
+
+    public bool Validate(BlittableJsonReaderObject obj, ErrorBuilder errorBuilder)
+    {
+        ObjectDisposedException.ThrowIf(_disposing.IsRaised(), nameof(SchemaValidator));
```
There is no `_disposing.Raise()` call so it will never be true? Is `_disposing` some leftover maybe?

haludi replied:
This was removed in the [21511](https://github.com/ravendb/ravendb/pull/21511)
The responsible for the resource is the caller.

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidator.cs:
```diff
@@ -0,0 +1,47 @@
+    }
+}
+
+public class RefSchema
```
Can we move it to a dedicated file?

haludi replied:
https://github.com/ravendb/ravendb/pull/21962

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/Validators/ElementSchemaRuleValidator.cs:
```diff
@@ -0,0 +1,171 @@
+    private bool IsOfRequiredType(object obj) => _typesRestriction == null || _typesRestriction.Length == 0 || _typesRestriction.Contains(obj?.GetType());
+}
+
+public static class ElementSchemaRuleValidatorFactory
```
Can we move it to a dedicated .cs file?

haludi replied:
https://github.com/ravendb/ravendb/pull/21962

arekpalinski replied:
great

arekpalinski commented on src/Raven.Server/Documents/SchemaValidation/Validators/ElementSchemaRuleValidator.cs:
```diff
@@ -0,0 +1,171 @@
+    public BlittableJsonReaderObject SchemaDefinition { get; init; }
+
+    // ReSharper disable once ConvertToPrimaryConstructor
+    public ElementSchemaRuleValidator(Type[] typesRestriction, ISchemaRuleValidator[] ruleValidators, SchemaPath schemaPath)
```
Can we have two separate constuctors to prevent from passing `null` as first argument - so we won't have code like `new ElementSchemaRuleValidator(null, [ifRequiredValidator], propertySchemaPath);`.

I think we could have:
- `public ElementSchemaRuleValidator(Type[] typesRestriction, ISchemaRuleValidator[] ruleValidators, SchemaPath schemaPath)` - where `typesRestriction` must not be null
- `public ElementSchemaRuleValidator(ISchemaRuleValidator[] ruleValidators, SchemaPath schemaPath)`

haludi replied:
https://github.com/ravendb/ravendb/pull/21962

Danielle9897 committed:
RavenDB-25524 Fix the sample schema

M4xymm committed:
RavenDB-25473 feat (schema validation): Enhanced progress tracking for schema validation operations with improved state handling (`isCompleted`, `isErrored`, `isLoading`). Refactored UI for test settings and validation monitoring.

grisha-kotler committed:
Merge branch 'v7.2' of https://github.com/ravendb/ravendb into v7.2

ppekrol committed:
Merge pull request #21951 from grisha-kotler/Schema-Validation-7.2

Merge 7.2 into feature Schema Validation

grisha-kotler committed:
RavenDB-25535 - make collection name validation case-insensitive

M4xymm committed:
RavenDB-25473 feat (schema validation): Refactored validation command and UI components. Improved error handling, state management, and introduced `Promise.allSettled` for async operations. Updated `DocumentSchemaPlayground` to better support form submission.

M4xymm committed:
RavenDB-25473 feat (schema validation): Replaced separate state flags (`isCompleted`, `isErrored`, `isLoading`) with a unified `status` field.

M4xymm committed:
RavenDB-24485 feat (operation types): Added `ValidateSchema` to the list of supported operation types.

M4xymm committed:
RavenDB-25473 feat (schema validation): Introduced dynamic table height calculation in `ValidationSchemaViewSheetPanel` using `virtualTableConstants`. Simplified column definition logic.

haludi committed:
RavenDB-22142 Dedicated files for RefSchema and ElementSchemaRuleValidatorFactory. Dedicated ctor for ElementSchemaRuleValidator without types restrictions

M4xymm committed:
RavenDB-25473 fix (document schema playground): Temporarily disable `DocumentSchemaPlaygroundAboutView` component rendering.

grisha-kotler committed:
RavenDB-22142 - initialize the SchemaValidatorCache only when we have enabled validators

ppekrol committed:
Merge branch 'v7.2' of https://github.com/ravendb/ravendb into feature/json-schema

gregolsky reviewed (commented):
1. Should we allow a single Validate operation run concurrently per database?
2.  Should we globally limit (using a configuration option) a number  of Validate operations run, considering hundred of dbs multitenant models?

3. From what I see validation code (SchemaValidationHelper) lacks logging at Info and Debug levels. Start (along with its parameters) and End of the operation should be logged. Debug messages should be logged at validation checkpoints/errors. Can we add this please?

4. Please add audit logging to handler's POST calls changing configuration and running validation.

grisha-kotler committed:
RavenDB-25530 - update the disabled status when the schema is unchanged and we only change the disabled status

grisha-kotler committed:
RavenDB-25545 - extend the GetReplicationOutgoingsFailureInfoCommand

grisha-kotler committed:
RavenDB-25545 - validate the documents according to the schema when it's coming from external replication

grisha-kotler committed:
RavenDB-25545 - change test class name

ppekrol committed:
Merge pull request #21969 from ppekrol/feature/json-schema

7.2 to feature/json-schema merge

grisha-kotler committed:
RavenDB-25587 - Skip schema validation for Refresh

grisha-kotler committed:
RavenDB-25587 - Skip schema validation for Data Archival

grisha-kotler committed:
RavenDB-25545 - fix failing tests

kalczur committed:
RavenDB-25473 Refactor rendering order of React providers in extensions class

grisha-kotler committed:
RavenDB-25545 - fix failing test

grisha-kotler committed:
RavenDB-25545 - use Any

M4xymm committed:
RavenDB-25575 feat (document schema playground): Prevent multiple sample schemas for the same collection and improve collection selection handling

grisha-kotler committed:
RavenDB-25545 - fix failing tests

grisha-kotler committed:
RavenDB-25545 - use a single flag

grisha-kotler committed:
RavenDB-25545 - add the IsInternal to the MergedDocumentReplicationCommandDto

grisha-kotler committed:
RavenDB-25564 - return the SchemaValidationConfiguration as the result of the command

grisha-kotler committed:
RavenDB-25564 - skip the schema validation for already accepted documents by the state machine

grisha-kotler committed:
RavenDB-25564 - validate the schema before saving the cluster-wide transactions

M4xymm committed:
RavenDB-25576 feat (document schema playground): Added collection filtering in `ValidationSchemaViewSheetPanel` with checkbox-based selection for playground mode. Enhanced validation UX with initial width configuration and improved test execution condition handling.

M4xymm committed:
RavenDB-25577 feat (schema validation): Added support for loading collection stats in `ValidationSchemaViewSheetPanel`. Introduced `asyncGetCollections` with state dispatch for collections tracking.

M4xymm committed:
RavenDB-25574 fix (document schema playground): Simplified layout by extracting `DocumentSchemaPlaygroundBody` as a separate component. Updated structure and removed redundant wrapper code.

haludi committed:
RavenDB-22142 Prevent resource exhaustion and log

	modified:   src/Raven.Server/Config/Categories/ConfigurationCategoryType.cs
	new file:   src/Raven.Server/Config/Categories/SchemaValidationConfiguration.cs
	modified:   src/Raven.Server/Config/RavenConfiguration.cs
	modified:   src/Raven.Server/Documents/DocumentDatabase.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForPost.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForGet.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Indexes/Index.cs
	modified:   src/Raven.Server/Documents/Indexes/SchemaValidation/IndexSchemaValidatorsByCollection.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/CurrentIndexingScope.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/ErrorBuilder.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ISchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/RefSchemas.cs
	new file:   src/Raven.Server/Documents/SchemaValidation/SchemaBuilderContext.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaRuleValidatorFactory.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaRuleValidatorFactoryHelper.cs
	new file:   src/Raven.Server/Documents/SchemaValidation/SchemaValidationContext.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidationHelper.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Array/ArraySchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Array/ContainsRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Array/UniqueItemsRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/ElementSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/ElementSchemaRuleValidatorFactory.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Number/MaximumSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Number/MinimumSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Number/MultipleOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/DependedRequiredSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/DependentSchemasSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/MaxPropertiesSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/MinPropertiesSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/ObjectSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/PropertyNamesSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/RequiredSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/String/MaximumLengthSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/String/MinimumLengthSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/String/PatternSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/AllOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/AnyOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/ConstantSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/EnumSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/GroupedIfThenElseSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/IfThenElseSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/MultiSubschemaAggregatorValidatorFactory.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/NotSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/OneOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/ReferenceSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs
	modified:   src/Raven.Server/ServerWide/ServerStore.cs
	modified:   test/FastTests/SchemaValidation/AdditionalPropertiesRulesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/AdvancedSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/ArraySchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/ConditionalSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/ConstantSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/GeneralSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/InvalidSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/NumberRulesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/PatternPropertiesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/PropertyNamesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/RefDefSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/SchemaValidationTestsBase.cs
	modified:   test/FastTests/SchemaValidation/StringRulesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/TypeSchemaValidationTests.cs
	modified:   test/SlowTests/SchemaValidation/BasicIntegrationSchemaValidationTests.cs
	new file:   test/SlowTests/SchemaValidation/SchemaValidationConfigurationTests.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs
	modified:   test/SlowTests/SlowTests.csproj

	modified:   src/Raven.Server/Config/Categories/ConfigurationCategoryType.cs
	new file:   src/Raven.Server/Config/Categories/SchemaValidationConfiguration.cs
	modified:   src/Raven.Server/Config/RavenConfiguration.cs
	modified:   src/Raven.Server/Documents/DocumentDatabase.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForPost.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForGet.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Indexes/Index.cs
	modified:   src/Raven.Server/Documents/Indexes/SchemaValidation/IndexSchemaValidatorsByCollection.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/CurrentIndexingScope.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/ErrorBuilder.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ISchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/RefSchemas.cs
	new file:   src/Raven.Server/Documents/SchemaValidation/SchemaBuilderContext.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaRuleValidatorFactory.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaRuleValidatorFactoryHelper.cs
	new file:   src/Raven.Server/Documents/SchemaValidation/SchemaValidationContext.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidationHelper.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Array/ArraySchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Array/ContainsRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Array/UniqueItemsRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/ElementSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/ElementSchemaRuleValidatorFactory.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Number/MaximumSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Number/MinimumSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Number/MultipleOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/DependedRequiredSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/DependentSchemasSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/MaxPropertiesSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/MinPropertiesSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/ObjectSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/PropertyNamesSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/RequiredSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/String/MaximumLengthSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/String/MinimumLengthSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/String/PatternSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/AllOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/AnyOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/ConstantSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/EnumSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/GroupedIfThenElseSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/IfThenElseSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/MultiSubschemaAggregatorValidatorFactory.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/NotSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/OneOfSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/ReferenceSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs
	modified:   src/Raven.Server/ServerWide/ServerStore.cs
	modified:   test/FastTests/SchemaValidation/AdditionalPropertiesRulesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/AdvancedSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/ArraySchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/ConditionalSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/ConstantSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/GeneralSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/InvalidSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/NumberRulesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/PatternPropertiesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/PropertyNamesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/RefDefSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/SchemaValidationTestsBase.cs
	modified:   test/FastTests/SchemaValidation/StringRulesSchemaValidationTests.cs
	modified:   test/FastTests/SchemaValidation/TypeSchemaValidationTests.cs
	modified:   test/SlowTests/SchemaValidation/BasicIntegrationSchemaValidationTests.cs
	new file:   test/SlowTests/SchemaValidation/SchemaValidationConfigurationTests.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs
	modified:   test/SlowTests/SlowTests.csproj

M4xymm committed:
RavenDB-25594 feat (schema definitions): Introduced support for schema definitions in index creation and editing

- Added `schemaDefinitionModel` for managing schema definitions.
- Updated index editor to support adding, formatting, and deleting schema definitions.
- Replaced the single `schemaDefinitionsText` field with an observable array of schemaDefinitionModel instances for better structure and management.

RavenDB-25594 feat (schema definitions): Enhanced schema editor with prettify JSON function and refined collection handling. Improved UI for schema definitions.

haludi commented:
> 1. Should we allow a single Validate operation run concurrently per database?

The operation runs with its own schema, so multiple operations can have different results.
Anyway, you can abort the operation.

> 2. Should we globally limit (using a configuration option) a number of Validate operations run, considering hundreds of DBs multitenant models?

I don't think this is more expensive than query.

> 3. From what I see validation code (SchemaValidationHelper) lacks logging at Info and Debug levels. Start (along with its parameters) and End of the operation should be logged. Debug messages should be logged at validation checkpoints/errors. Can we add this please?

I will

> 4. Please add audit logging to handler's POST calls changing configuration and running validation.

Ok

M4xymm committed:
RavenDB-25576 fix (schema validation): Replaced `hideSelection` with `showSelection` for consistency in `ValidationSchemaViewSheetPanel` and related components

M4xymm committed:
RavenDB-25594 refactor (schema definitions): Simplify schema handling and improve JSON formatting

- Removed unused `toDto` method in `schemaDefinitionModel`.
- Renamed `prettifyContent` to `prettifyJson` for clarity.
- Updated index definitions to use default values for `schemaDefinitions`.
- Refined schema formatting logic to reduce redundancy.

M4xymm committed:
RavenDB-25576 refactor (schema validation): Unified `activeValidators` logic and initialized `selectedCollections` state to improve readability and efficiency in `ValidationSchemaViewSheetPanel`.

grisha-kotler committed:
RavenDB-25564 - prevent NRE

haludi committed:
RavenDB-22142 Removing AuditLog from get, fix spaces and config default

grisha-kotler committed:
RavenDB-25627 - use a separate context for each SchemaValidatorCache which isn't allocated from the context pool

grisha-kotler committed:
RavenDB-25627 - publish the new SchemaValidatorCache only after it's updated

haludi committed:
RavenDB-22142 Remove timeout for document validation

haludi committed:
RavenDB-25627 Avoid reading Blittalbe object concurrently

	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidationHelper.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/ConstantSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/EnumSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Untyped/FixedValueSchemaRuleValidator.cs
	modified:   src/Raven.Server/Properties/launchSettings.json
	modified:   src/Sparrow/Json/BlittableJsonReaderArray.cs
	modified:   src/Sparrow/Json/BlittableJsonReaderObject.cs
	modified:   test/SlowTests/SchemaValidation/BasicIntegrationSchemaValidationTests.cs
	new file:   test/SlowTests/SchemaValidation/SchemaValidationSlowUnitTests.cs
	modified:   test/SlowTests/SlowTests.csproj

ppekrol committed:
Merge branch 'v7.2' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #22015 from ppekrol/feature/json-schema

Merge 7.2 into feature/json-schema

haludi committed:
RavenDB-25627 CloneForConcurrentRead also for string values

haludi committed:
RavenDB-25557 Skip schema validation for documents during import

haludi committed:
RavenDB-25557 Don't validate for restore

haludi committed:
RavenDB-25543 Avoid error message with empty path

ppekrol committed:
Merge branch 'v7.2' of https://github.com/ravendb/ravendb into feature/json-schema

ppekrol committed:
Merge pull request #22046 from ppekrol/feature/json-schema

7.2 into feature/json-schema merge

haludi committed:
RavenDB-22142 Catch only SchemaValidationException in import

grisha-kotler committed:
RavenDB-22142 - remove the finalizer
